### DISCUSSION TITLE
Swagger UI for Nuxt

### DISCUSSION THREAD

**Discussion by BracketJohn** [nuxt/nuxt#16165] [Questions]
Hey there!

We are currently moving from a python + nuxt2 approach to our projects to a nuxt3 fullstack approach. This is all great and has many advantages, but crucially it also lacks some features that really shined in our own setup (as would be expected).

With python we used [fastapi](https://github.com/tiangolo/fastapi), it has many great things out of the box, one of them being a Swagger UI for the backend: 
![image](https://user-images.githubusercontent.com/25911230/165047196-91f3a9cc-2bfc-47aa-90d3-c24b608ad83f.png)

As far as I know this does not exist in nuxt3 or it's backend `h3` implementation. But, it is a great feature for newcomers, product managers and smoke-testing (: 

Does anyone have an idea on how to set this up? Would this be a good feature / module / plugin / ...? Ideally, it would use TypeScript to infer the required openAPI schema automagically.

Thanks!

edit: we will attempt to do this oursevles as a nuxt module to include in [sidebase](https://github.com/sidebase/sidebase)

**Comment by drewdru** [comment]
I found a way to generate swagger documentation for Nuxt 3 but only with docstrings:

1. Install `swagger-ui`
I tried to install `swagger-ui` with npm but I had some troubles with it therefore I imported it as external script:

components/Swagger.vue
```
<template>
  <div id="swagger-ui"></div>
</template>

<script lang="ts" setup>
useMeta({
  script: [
    {
      type: "text/javascript",
      src: "https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.11.1/swagger-ui-bundle.min.js",
      async: true,
    },
    {
      type: "text/javascript",
      src: "https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.11.1/swagger-ui-standalone-preset.min.js",
      async: true,
    },
  ],
  link: [
    {
      rel: "stylesheet",
      href: "https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.11.1/swagger-ui.min.css",
    },
  ],
});
if (process.client) {
  const config = useRuntimeConfig();
  const ui = SwaggerUIBundle({
    url: `//${config.VITE_DOMAIN_NAME}/api/docs`,
    dom_id: "#swagger-ui",
    deepLinking: true,
    presets: [SwaggerUIBundle.presets.apis, SwaggerUIStandalonePreset],
    layout: "StandaloneLayout",
  });
  window.ui = ui;
}
</script>
```

2. Install `swagger-jsdoc`
```
npm install swagger-jsdoc
```

3. Generate swagger-jsdoc on some endpoint

server/api/docs.ts
```
import type { IncomingMessage, ServerResponse } from "http";
import swaggerJSDoc from "swagger-jsdoc";

export default (_req: IncomingMessage, _res: ServerResponse) => {
  const swaggerDefinition = {
    openapi: "3.0.0",
    info: {
      title: "TITLE",
      version: "1.0.0",
    },
    servers: [{ url: "/api" }],
    schemes:
      process.env.SWAGGER_SCHEMA_HTTPS === "true"
        ? ["https"]
        : ["http", "https"],
    components: {
      securitySchemes: {
        BearerAuth: {
          type: "apiKey",
          name: "Authorization",
          in: "header",
        },
      },
    },
    security: {
      ApiKeyAuth: [],
      OAuth2: {},
    },
  };

  const options = {
    swaggerDefinition,
    apis: ["server/api/**/*.ts"],
  };

  return swaggerJSDoc(options);
};
```
5. And now you can add swagger docstrings to your endpoint

server/api/hello.ts
```
import type { IncomingMessage, ServerResponse } from "http";
import { useQuery } from "h3";
/**
 * @openapi
 * /hello:
 *   get:
 *     description: Welcome to swagger-jsdoc!
 *     responses:
 *       200:
 *         description: Returns a mysterious object.
 */
export default (req: IncomingMessage, _res: ServerResponse) => {
  return {
    query: useQuery(req),
  };
};
```

**Reply by BracketJohn** [reply]
This is great, thanks!

I was however looking for something more interactive, playground style! In the above playground that I screenshotted you can call the endpoints, authorize and un-authorize yourself, see documentation and let the product manager try out everything without having to look into code! It's really neat and I would love to see nuxt3 have it (:

**Reply by drewdru** [reply]
If you add component components/Swagger.vue on any page you will get what you want.

**Comment by pi0** [comment]
Hi! Here is open issue to introduce devtools and swagger UI: https://github.com/unjs/nitro/issues/4

**Reply by BracketJohn** [reply]
Wow very cool, thanks! 

IMHO it would be a great to have feature / module for the nuxt ecosystem, as it helps product managers and developers alike (: Thanks for following up on this request!

I understand that https://github.com/unjs/nitro/issues/4 will introduce the server-side compabilities, so I'd leave this discussion open for the frontend related side -> does that make sense?

**Comment by justindasilva** [comment]
All of a sudden, I'm getting an error (This was working a few weeks back):
```
[nitro] [dev] [uncaughtException] ReferenceError: window is not defined
    at Proxy.bn (/docs/node_modules/swagger-ui-dist/swagger-ui-bundle.js:2:1054382)
    at Timeout._onTimeout (/docs/components/content/Swagger.vue:12:29)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
```

I have this code:
```
<script setup lang="ts">
import { SwaggerUIBundle } from "swagger-ui-dist";
import swagger from '../../.build/swagger.json'

setTimeout(() => {
  SwaggerUIBundle({
    dom_id: "#swagger-ui",
    spec: swagger,
  })
})
</script>

<template>
  <div id="swagger-ui">Coming soon</div>
</template>

<style>
  @import "swagger-ui-dist/swagger-ui.css";
</style>
```

**Reply by ram-you** [reply]
@justindasilva Change setTimeout to onMounted
```js
onMounted(() => {
  SwaggerUIBundle({
    dom_id: "#swagger-ui",
    spec: swagger,
  })
})
```

**Reply by justindasilva** [reply]
I gave that a shot as seen below. 
![Screen Shot 2022-09-12 at 3 24 52 PM](https://user-images.githubusercontent.com/882534/189738743-766bad35-6505-47ed-b2bd-594290597c67.png)

So if I go to http://localhost:3001/developers/api, my Swagger documentation does show up properly.

But if I am trying to access the route from another route, I get the following error (the Swagger does not render in this case):

![Screen Shot 2022-09-12 at 3 27 05 PM](https://user-images.githubusercontent.com/882534/189739045-d447e85a-813a-4ab5-83ab-628b850f81be.png)

**Reply by ram-you** [reply]
Hi, try to use **domNode** in place of **dom_id**
Add a **console.log(swagger)** to verify data content

``` js
<script setup>
import { SwaggerUIBundle } from "swagger-ui-dist";
import swagger from '../../.build/swagger.json'
console.log(swagger) 

const swaggerDomNode = ref(null);

onMounted(() => {
  SwaggerUIBundle({
    domNode: swaggerDomNode.value,
    spec: swagger || {},
  })
})

</script>
<template>
  <div ref="swaggerDomNode"></div>
</template>

<style>
@import "swagger-ui-dist/swagger-ui.css";
</style>
```

**Reply by justindasilva** [reply]
Wow, that worked 🎉 

Thank you @ram-you

**Comment by 8bu** [comment]
https://www.npmjs.com/package/next-swagger-doc
It will be great if we have something similar in NuxtJS, since Nextjs already have one.

**Reply by BracketJohn** [reply]
Agreed! I think we'll create one at sidebase: https://github.com/sidebase (:

**Comment by BracketJohn** [comment]
Update for the thread: As we actively started to contribute to the ecosystem with sidebase (https://github.com/sidebase) this module is now on our list for implementation. So hopefully I can share a first version here soon (:

**Reply by justindasilva** [reply]
I look forward to it!
